### DISCUSSION TITLE
Using RadioLib modules in generics

### DISCUSSION THREAD

**Discussion by phretor** [jgromes/RadioLib#618] [Q&A] [answered]
Hi!

I think that if there's one person who can understand this intricate design, that person is @jgromes 😁 

I'm wrapping RadioLib in a weird way in a so-called `RadioLibWrapper<T>` which is basically a generic interface class that "mandates" which methods must be implemented by all subclasses (e.g., `setFrequency`, `setCrcFiltering`), by declaring such methods as `virtual` and by returning an error code in their default implementation.

Example (full code [here](https://github.com/rfquack/RFQuack/blob/master/src/radio/RadioLibWrapper.h)):

```cpp
class RadioLibWrapper {
public:
   // ...

    virtual int16_t begin() {
      return T::begin();
    }

   //...

    virtual int16_t setFrequency(float freq) {
      Log.error(F("setFrequency was not implemented."));

      return ERR_COMMAND_NOT_IMPLEMENTED;
    }

   //...

    virtual int16_t setFrequencyDeviation(float freqDev) override {
      Log.error(F("setFrequencyDeviation was not implemented."));

      return ERR_COMMAND_NOT_IMPLEMENTED;
    }
    
    //...

    virtual int16_t setCrcFiltering(bool crcOn) {
      Log.error(F("setCrcFiltering was not implemented."));

      return ERR_COMMAND_NOT_IMPLEMENTED;
    }

};
```

For each of the wrapped RadioLib modules I then declare a subclass, which proxies most of the methods and overrides some others. The proxied ones are interesting:

```cpp

// here, RF69 is from RadioLib's RF69.h

class RFQRF69 : public RadioLibWrapper<RF69> {
public:
    using RF69::setPreambleLength;
    using RF69::setOutputPower;
    using RF69::setPromiscuousMode;
    using RF69::variablePacketLengthMode;
    using RF69::setCrcFiltering;
    using RF69::setRxBandwidth;
    using RF69::fixedPacketLengthMode;
    using RF69::setFrequencyDeviation;
    using RF69::getFrequencyDeviation;
    using RF69::setBitRate;

   //...see, most of the methods are proxied up here 👆

    int16_t begin() override {
      int16_t state = RadioLibWrapper::begin();     // note that this in turns calls T::begin() above 👆

      //..
    }

```

Now, when I call `RFQRF69::begin()` I expect that `RadioLibWrapper::begin()` is called inside it, which in turn calls `RF69::begin()`. All of these calls happen.

Now, when `RF69::begin()` calls its `setFrequency()` I **don't** expect that `RadioLibWrapper::setFrequency()` gets called, but unfortunately that's what happens.

If I comment out `virtual RadioLibWrapper::setFrequency() {}` implementation, the `RF69::setFrequency()` symbol is called as expected.

Should I maybe call `T::begin()` inside `RFQ69::begin()`?

Full explanation with repro: https://community.platformio.org/t/linking-order-problem-with-virtual-override-methods/30584

At this point I don't know if it's a linking problem or a bad inheritance design. Any hint would be appreciated.

**Comment by jgromes** [comment, marked as answer]
I tried to write a minimum example in [cpp.sh](https://cpp.sh/), here it is:

```c++
#include <iostream>

template<typename T>
class Wrapper: protected T {
    public:
        virtual int foo() {
            std::cout << "Wrapper:foo()\n";
            return(T::foo());  
        }
        
        virtual int boo() {
            std::cout << "Wrapper:boo()\n";
            return(T::boo());  
        }
};

class InternalClass {
   public:
        int foo() {
            std::cout << "InternalClass:foo()\n";
            boo();
            return(0);
        }
        
        int boo() {
            std::cout << "InternalClass:boo()\n";
            return(1);
        }
};

class WrappedClass: public Wrapper<InternalClass> {
    using InternalClass::boo;
    
    public:
        int foo() override {
            std::cout << "WrappedClass:foo()\n";
            return(Wrapper::foo());
        }
};

int main()
{
   WrappedClass wrapped;
   int res = wrapped.foo();
   std::cout << res;
}
```

This is the output:

```
WrappedClass:foo()
Wrapper:foo()
InternalClass:foo()
InternalClass:boo()
0 
```

It seems to be behaving as expected, i.e. `InternalClass:boo()` is called, as opposed to `Wrapper::boo()`. Could you check I reproduced your class structure correctly?

**Reply by phretor** [reply]
It is behaving as expected and it reflects the case. Could a wrong linking order mess up with this, given that the overridden method is `virtual`?

**Reply by jgromes** [reply]
Not sure, what are the linker options?
